packages/mattermost-redux/src/selectors/entities/channels.ts

```diff
@@ -879,6 +921,105 @@ export const getUnreadChannels: (state: GlobalState, lastUnreadChannel?: Channel
     },
 );
 
+function maxDefined(a: number, b?: number) {
+    return typeof b === 'undefined' ? a : Math.max(a, b);
+}
+
+type LastUnreadChannel = (Channel & {hadMentions: boolean});
+
+export const getUnsortedAllTeamsUnreadChannels: (state: GlobalState) => Channel[] = createSelector(
+    'getAllTeamsUnreadChannels',
+    getCurrentUser,
+    getUsers,
+    getUserIdsInChannels,
+    getAllChannels,
+    getAllTeamsUnreadChannelIds,
+    getTeammateNameDisplaySetting,
+    (currentUser, profiles, userIdsInChannels, channels, allTeamsUnreadChannelIds, settings) => {
+        // If we receive an unread for a channel and then a mention the channel
+        // won't be sorted correctly until we receive a message in another channel
+        if (!currentUser) {
+            return [];
+        }
+
+        return allTeamsUnreadChannelIds.filter((id) => channels[id] && channels[id].delete_at === 0).map((id) => {
+            const c = channels[id];
+
+            if (c.type === General.DM_CHANNEL || c.type === General.GM_CHANNEL) {
+                return completeDirectChannelDisplayName(currentUser.id, profiles, userIdsInChannels[id], settings!, c);
+            }
+
+            return c;
+        });
+    },
+);
+
+const sortChannels: (state: GlobalState, lastUnreadChannel: LastUnreadChannel | null) => Channel[] = createSelector(
```

Comment:
@hmhealey Could you please take a look at this I am not sure if we should move code here https://github.com/mattermost/mattermost-webapp/blob/be0aa1931b395c54ac53ec8624ef9811bc9faadb/selectors/views/channel_sidebar.ts#L142 as the method I have added is just duplicate of it.

Reply:
I think we could share the result function (lines 964-1014), but we can't share the rest of this because one uses `lastUnreadChannel` and the other doesn't. Since one place passes in a value for it and the other does not, those two calls will end up overwriting the result from the previous one, causing it to recalculate constantly. 

What we could do though is share sorting logic by sharing the result function between the two so that it looks like:
```typescript
// mattermost-webapp/selectors/views/channel_sidebar.ts
    const sortChannels = createSelector(
        'sortChannels',
        (state: GlobalState, channels: Channel[]) => channels,
        getMyChannelMemberships,
        getLastPostPerChannel,
        (state: GlobalState) => state.views.channel.lastUnreadChannel,
        isCollapsedThreadsEnabled,
        (channels, myMembers, lastPosts, lastUnreadChannel, crtEnabled) => {
            return sortUnreadChannels(channels, myMembers, lastPosts, lastUnreadChannel, crtEnabled);
        },
    );

// packages/mattermost-redux/src/selectors/entities/channels.ts
const getSortedAllTeamsUnreadChannels = createSelector(
    'getSortedAllTeamsUnreadChannels',
    getUnsortedAllTeamsUnreadChannels,
    getMyChannelMemberships,
    getLastPostPerChannel,
    isCollapsedThreadsEnabled,
    (channels, myMembers, lastPosts, crtEnabled) => {
        return sortUnreadChannels(channels, myMembers, lastPosts, null, crtEnabled);
    },
);
```

And then sortUnreadChannels is shared between the two. You can probably move that code here then to avoid an import loop between these two files.